rust/main/chains/hyperlane-ethereum/src/signer/yubi.rs

```diff
@@ -0,0 +1,146 @@
+/// too quickly, the devices resources will become exhausted leading to failures.
+static YUBIHSM_CLIENT_CACHE: LazyLock<Mutex<HashMap<YubihsmClientCacheKey, WrappedSigner>>> =
+    LazyLock::new(|| Mutex::new(HashMap::new()));
+
+/// Get a signer.
+/// If a matching signer is in the cache, the cached version is returned. Otherwise a new signer is initialized, and added to the cache.
+/// This function is thread safe.
+fn get_signer_with_cache(
+    http_config: HttpConfig,
+    credentials: Credentials,
+    id: object::Id,
+) -> WrappedSigner {
+    let cache_key = YubihsmClientCacheKey {
+        config: http_config.clone(),
+        credentials: credentials.clone(),
+        id: id.clone(),
+    };
+    let mut cache = YUBIHSM_CLIENT_CACHE.lock().unwrap();
+
+    // If a client already exists for this key, return a clone of the Arc.
+    if let Some(existing_client) = cache.get(&cache_key) {
+        return Arc::clone(existing_client);
+    }
+
+    // Build a new client and add it to the cache.
+    let client = YubiHsmSigner::connect(http_config, credentials);
+    let signer = Arc::new(YubiSigner::create(client, id).unwrap());
+    cache.insert(cache_key, Arc::clone(&signer));
+
+    signer
+}
+
+/// A wallet connected to a yubihsm2
+pub type YubiHsmWallet = Wallet<YubiHsmSigner>;
+
+/// The signer that is wrapped in a thread safe manner.
+type WrappedSigner = Arc<yubihsm::ecdsa::Signer<ethers_core::k256::Secp256k1>>;
+
+/// YubiHsmSigner is a signer that is backed by a yubihsm2.
+/// It wraps a normal signer in Arc, while implementing Signer for Wallet, such that a Wallet holding a YubiHsmSigner is cloneable.
+#[derive(Clone)]
+pub struct YubiHsmSigner {
+    wrapped: WrappedSigner,
+}
+
+impl YubiHsmSigner {
+    /// Connect to the yubihsm and return a wallet backed by the connection
+    /// Utilizes the global cache.
+    pub fn new_wallet(
+        http_config: HttpConfig,
+        credentials: Credentials,
+        id: object::Id,
+    ) -> Wallet<Self> {
+        let signer = YubiHsmSigner::new(http_config, credentials, id);
+        let address = signer.get_address();
+
+        Wallet::new_with_signer(signer, address, 1)
+    }
+
+    /// Connect to the yubihsm and return a signer
+    pub fn new(http_config: HttpConfig, credentials: Credentials, id: object::Id) -> Self {
+        let wrapped = get_signer_with_cache(http_config, credentials, id);
+
+        Self { wrapped }
+    }
+
+    /// Performs initial connection to the device and returns an initialized signer.
+    /// This function does not use a cache. Care should be taken to not exhaust the number of sessions on the yubihsm when calling this function.
+    fn connect(http_config: HttpConfig, credentials: Credentials) -> Client {
+        let connector = ethers::signers::yubihsm::Connector::http(&http_config);
+        Client::open(connector.clone(), credentials.clone(), true).unwrap()
```

Comment:
I wonder if we want to return a `Result` here and let user of the method to decide what to do in the case of an error?